content/en/blog/_posts/TBD-admission-controllers-for-container-drift/index.md

```diff
@@ -0,0 +1,90 @@
+
+### 1. Admit interactive kubectl command requests
+First of all, we need to enable a validating webhook that sends qualified requests to our admission controller service. In the interest of interactive kubectl commands, we configure the webhook’s rules with resources as "[pods/exec, pods/attach]", and operations as “CONNECT”. They tell the K8s API server that all “exec” and "attach" requests should be subject to our admission controller service. We also specify a "service" reference (or "url" that gives the location of the webhook) and “caBundle” to provide its certificate, both under the “clientConfig” stanza. Here is a short example of what our ValidatingWebhookConfiguration object looks like:
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: example-validating-webhook-config
+webhooks:
+  - name: validate-pod-interaction.example.com
+    sideEffects: None
+    rules:
+      - apiGroups: ["*"]
+        apiVersions: ["*"]
+        operations: ["CONNECT"]
+        resources: ["pods/exec", "pods/attach"]
+    failurePolicy: Fail
+    clientConfig:
+      service:
+        name: example-service
+        namespace: kube-exec-controller
+        path: "/admit-pod-interaction"
+      caBundle: "{{CABUNDLE_VALUE}}"
+    admissionReviewVersions: ["v1", "v1beta1"]
+```
+
+### 2. Tag the pod with potentially mutated containers
+Once a request of `kubectl exec` comes to our admission controller service, it tags the target pod by adding some customized K8s labels. In this way, we can not only query all the pods with potentially mutated containers, but also enable our service to retrieve previously identified pods in case itself gets restarted. The service cannot directly label the pod and return the updated object in its admission response. This is because the object stored in these admission requests is of type “PodExecOptions”, a child object of type “Pod”. As a result, there is a separate process in our admission controller service that patches the labels. The service returns an admission response as allowed after passing that pod reference to its K8s client, who is responsible for labeling and sending related K8s events to the target pod. Developers can check whether their pods are affected or not, simply by looking at its labels or running a `kubectl describe` command:
```

Comment:
```suggestion
Once a request of `kubectl exec` comes to our admission controller service, the admission controller makes an internal note to label the associated Pod.The added labels mean that we can not only query all the pods with potentially mutated containers, but also enable the security mechanism to retrieve previously id. That labelling happens asynchronously. entified pods in case itself gets restarted.

The admission control cannot directly modify the targeted in its admission response. This is because the `pods/exec` request is against a subresource of the Pod API and the API kind for that subresource is `PodExecOptions`. As a result, there is a separate process in our admission controller code that patches the labels asynchronously. The admission controller allays permits the exec, then acts as a client of the Kubernetes API to label the target Pod and to log related events. Developers can check whether their pods are affected or not using `kubectl` or similar tools. For example:
```

Reply:
(API design note: `box.com/podInitialInteractionTimestamp=1632524400`, `box.com/podInteractorUsername=username-1` and `box.com/podTTLDuration=1h0m0s` should ideally be _annotations_, and there should be a separate label to tell clients to look for those 3 annotations).

If these actually are annotations, let's make that clear.

Reply:
I think we specifically selected _labels_ here in our API design because all these values would be used to identify affected pods (e.g. filtered by given username, interaction start time, or TTL in different environments of clusters). [The K8s docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) also seem to align with this kind of usage of _labels_:

> Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users... Non-identifying information should be recorded using annotations.

As pointed out by the doc above, we used _annotations_ for the non-identifying and mutable info (such as requested extension time from users). This also made our admission controller service easier when admitting pods' metadata change, as it doesn't allow any updates to the above _labels_ but supports updating the values of _annotations_.

Reply:
Addressed in [this commit](https://github.com/kubernetes/website/pull/29761/commits/540390b811ac7f193b8cc62d71f4647db8999dd0).